src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_wrapper_program.cpp

```diff
@@ -0,0 +1,1068 @@
+                     id2string(wrapper_symbol.name),
+                     "__no_alloc_dealloc_in_requires",
+                     source_locationt{}, // TODO add property class and comment
+                     wrapper_symbol.mode,
+                     goto_model.symbol_table)
+                     .symbol_expr();
+
+  postamble.add(goto_programt::make_decl(check_var));
+
+  postamble.add(goto_programt::make_function_call(code_function_callt{
+    check_var,
+    library
+      .dfcc_fun_symbol
+        [dfcc_funt::WRITE_SET_CHECK_ALLOCATED_DEALLOCATED_IS_EMPTY]
+      .symbol_expr(),
+    {address_of_write_set}}));
+
+  // TODO add property class on assertion source_location
+  source_locationt sl;
+  sl.set_function(wrapper_symbol.name);
+  sl.set_comment("Check that requires do not allocate or deallocate memory");
+  postamble.add(goto_programt::make_assertion(check_var, sl));
+  postamble.add(goto_programt::make_dead(check_var));
+
+  // generate write set release and DEAD instructions
+  {
+    code_function_callt call(
+      library.dfcc_fun_symbol.at(dfcc_funt::WRITE_SET_RELEASE).symbol_expr(),
+      {address_of_write_set});
+    postamble.add(goto_programt::make_function_call(call));
+    postamble.add(goto_programt::make_dead(write_set));
+  }
+  log.debug() << "<-dfcc_dsl_wrapper_programt::encode_requires_write_set()"
+              << messaget::eom;
+}
+
+void dfcc_dsl_wrapper_programt::encode_ensures_write_set()
+{
+  log.debug() << "->dfcc_dsl_wrapper_programt::encode_ensures_write_set()"
+              << messaget::eom;
+
+  PRECONDITION(ensures_write_set_symbol_opt);
+  PRECONDITION(addr_of_ensures_write_set_symbol_opt);
+
+  // call set_create(
+  //   ensures_write_set,
+  //   assigns_size = 0,
+  //   frees_size = 0,
+  //   replacement_mode = false,
+  //   assume_requires_ctx = false,
+  //   assert_requires_ctx = false,
+  //   assume_ensures_ctx = contract_mode != check,
+  //   assert_ensures_ctx = contract_mode == check,
+  // )
+  const auto write_set = ensures_write_set_symbol_opt->symbol_expr();
+  preamble.add(goto_programt::make_decl(write_set));
+
+  const auto address_of_write_set =
+    addr_of_ensures_write_set_symbol_opt->symbol_expr();
+  preamble.add(goto_programt::make_decl(address_of_write_set));
+  preamble.add(goto_programt::make_assignment(
+    address_of_write_set, address_of_exprt(write_set)));
+
+  auto function_symbol =
+    library.dfcc_fun_symbol.at(dfcc_funt::WRITE_SET_CREATE).symbol_expr();
+  code_function_callt call(function_symbol);
+  auto &arguments = call.arguments();
+
+  // write set
+  arguments.emplace_back(address_of_write_set);
+
+  // max assigns clause size
+  arguments.emplace_back(from_integer(0, size_type()));
+
+  // max frees clause size
+  arguments.emplace_back(from_integer(0, size_type()));
+
+  // replacement mode
+  arguments.emplace_back((exprt)false_exprt());
+
+  // assume_requires_ctx
+  arguments.emplace_back((exprt)false_exprt());
```

Comment:
I'm not sure I understand why this is repeated.  Are we trying to match certain fields by installing these in order?  Is there a better way to do this?